// @sanity/studio-secrets/lib/index.mjs
import { jsx, jsxs } from "react/jsx-runtime";
import { Dialog, Card, Stack, Button, Text, TextInput } from "@sanity/ui";
import { useState, useEffect, useCallback } from "react";
import { useClient } from "sanity";
const query = "* [_id == $id] {secrets}[0]", type = "pluginSecrets";
function useSecrets(namespace) {
  const [loading, setLoading] = useState(!0), [secrets, setSecrets] = useState(), client = useClient({ apiVersion: "2021-03-01" }), id = `secrets.${namespace}`;
  return useEffect(() => {
    const subscription = client.observable.listen(query, { id }, { visibility: "query", tag: "secrets.listen" }).subscribe((result) => {
      setSecrets(result?.result?.secrets);
    });
    return () => {
      subscription.unsubscribe();
    };
  }, [id, client]), useEffect(() => {
    async function fetchData() {
      client.fetch(query, { id }, { tag: "secrets.get" }).then((doc) => setSecrets(doc?.secrets)).finally(() => setLoading(!1));
    }
    fetchData();
  }, [id, client]), { loading, secrets, storeSecrets: (updatedSecret) => {
    setLoading(!0);
    const keysPatch = client.patch(id).set({ secrets: updatedSecret });
    client.transaction().createIfNotExists({ _id: id, _type: type }).patch(keysPatch).commit({ visibility: "async", tag: "secrets.store" }).finally(() => setLoading(!1));
  } };
}
const SettingsView = ({
  namespace,
  keys,
  onClose,
  title = "Configure"
}) => {
  const { loading, secrets, storeSecrets } = useSecrets(namespace), [newSecrets, setNewSecrets] = useState({});
  useEffect(() => {
    secrets && setNewSecrets(secrets);
  }, [secrets]);
  const onClick = useCallback(() => storeSecrets(newSecrets), [storeSecrets, newSecrets]);
  return /* @__PURE__ */ jsx(Dialog, { animate: !0, id: "translation-settings", onClose, header: title, children: /* @__PURE__ */ jsx(Card, { padding: 3, children: /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
    keys.map((keyEntry) => /* @__PURE__ */ jsx(
      SettingsKeyEntry,
      {
        keyEntry,
        loading,
        newSecrets,
        setNewSecrets
      },
      keyEntry.key
    )),
    /* @__PURE__ */ jsx(
      Button,
      {
        disabled: loading,
        onClick,
        text: loading ? "Loading\u2026" : "Save",
        tone: "positive"
      }
    )
  ] }) }) });
};
function SettingsKeyEntry({ loading, setNewSecrets, newSecrets, keyEntry }) {
  const onChange = useCallback(
    (event) => {
      const target = event.currentTarget, { value } = target;
      setNewSecrets((prevState) => {
        const newState = { ...prevState };
        return newState[keyEntry.key] = value, newState;
      });
    },
    [keyEntry, setNewSecrets]
  );
  return /* @__PURE__ */ jsxs(Stack, { space: 2, children: [
    /* @__PURE__ */ jsx(Text, { as: "label", weight: "semibold", size: 1, children: keyEntry.title }),
    keyEntry.description && /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: keyEntry.description }),
    /* @__PURE__ */ jsx(TextInput, { disabled: loading, onChange, value: newSecrets[keyEntry.key] ?? "" })
  ] });
}
export {
  SettingsView,
  useSecrets
};
