// @sanity/studio-secrets/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var jsxRuntime = require("react/jsx-runtime"), ui = require("@sanity/ui"), react = require("react"), sanity = require("sanity");
const query = "* [_id == $id] {secrets}[0]", type = "pluginSecrets";
function useSecrets(namespace) {
  const [loading, setLoading] = react.useState(!0), [secrets, setSecrets] = react.useState(), client = sanity.useClient({ apiVersion: "2021-03-01" }), id = `secrets.${namespace}`;
  return react.useEffect(() => {
    const subscription = client.observable.listen(query, { id }, { visibility: "query", tag: "secrets.listen" }).subscribe((result) => {
      setSecrets(result?.result?.secrets);
    });
    return () => {
      subscription.unsubscribe();
    };
  }, [id, client]), react.useEffect(() => {
    async function fetchData() {
      client.fetch(query, { id }, { tag: "secrets.get" }).then((doc) => setSecrets(doc?.secrets)).finally(() => setLoading(!1));
    }
    fetchData();
  }, [id, client]), { loading, secrets, storeSecrets: (updatedSecret) => {
    setLoading(!0);
    const keysPatch = client.patch(id).set({ secrets: updatedSecret });
    client.transaction().createIfNotExists({ _id: id, _type: type }).patch(keysPatch).commit({ visibility: "async", tag: "secrets.store" }).finally(() => setLoading(!1));
  } };
}
const SettingsView = ({
  namespace,
  keys,
  onClose,
  title = "Configure"
}) => {
  const { loading, secrets, storeSecrets } = useSecrets(namespace), [newSecrets, setNewSecrets] = react.useState({});
  react.useEffect(() => {
    secrets && setNewSecrets(secrets);
  }, [secrets]);
  const onClick = react.useCallback(() => storeSecrets(newSecrets), [storeSecrets, newSecrets]);
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Dialog, { animate: !0, id: "translation-settings", onClose, header: title, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { padding: 3, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 3, children: [
    keys.map((keyEntry) => /* @__PURE__ */ jsxRuntime.jsx(
      SettingsKeyEntry,
      {
        keyEntry,
        loading,
        newSecrets,
        setNewSecrets
      },
      keyEntry.key
    )),
    /* @__PURE__ */ jsxRuntime.jsx(
      ui.Button,
      {
        disabled: loading,
        onClick,
        text: loading ? "Loading\u2026" : "Save",
        tone: "positive"
      }
    )
  ] }) }) });
};
function SettingsKeyEntry({ loading, setNewSecrets, newSecrets, keyEntry }) {
  const onChange = react.useCallback(
    (event) => {
      const target = event.currentTarget, { value } = target;
      setNewSecrets((prevState) => {
        const newState = { ...prevState };
        return newState[keyEntry.key] = value, newState;
      });
    },
    [keyEntry, setNewSecrets]
  );
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 2, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { as: "label", weight: "semibold", size: 1, children: keyEntry.title }),
    keyEntry.description && /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { muted: !0, size: 1, children: keyEntry.description }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.TextInput, { disabled: loading, onChange, value: newSecrets[keyEntry.key] ?? "" })
  ] });
}
exports.SettingsView = SettingsView;
exports.useSecrets = useSecrets;
